### PR TITLE
hwdb: Add ID_INPUT_JOYSTICK_INTEGRATION property

### PR DESCRIPTION
**Pull request by hadess**
When the joystick is integrated directly into the machine, knowing
that the device is internal allows us to disable attached functionality
when the device is not used or inaccessible.

For example, this allows disabling rumble and accelerometer on
flip-console-like devices like the GPD-XD.

### CODE REVIEW

**Comment by hadess** [comment]
CC: @whot @Kekun

File: rules/70-joystick.rules [hunk 1]
```diff
@@ -0,0 +1,13 @@
+# do not edit this file, it will be overwritten on update
+
+ACTION=="remove", GOTO="joystick_end"
+ENV{ID_INPUT}=="", GOTO="joystick_end"
+ENV{ID_INPUT_TOUCHPAD}=="", GOTO="joystick_end"
```
Comment (poettering) [inline rules/70-joystick.rules]: ID_INPUT_TOUCHPAD? shouldn't this be ID_INPUT_JOYSTICK?

Comment (hadess) [reply rules/70-joystick.rules]: Completely. Can you tell I don't have the hardware and failed the copy-pasta :/

**Comment by hadess** [comment]
I CC:'ed Peter to look for sanity checking, and Adrien to actually test.

I guess @jwrdegoede could also test.

**Comment by keszybz** [comment]
lgtm too. No hardware to test, I haven't seen a joystick in years ;)

**Comment by Kekun** [comment]
I have the hardware, with some guidance I could try the patch tomorrow if needed.

File: hwdb/70-joystick.hwdb [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+# If the property is missing, user-space can assume
+# ID_INPUT_JOYSTICK_INTEGRATION=external
+
+###########################################################
+# GPD
+###########################################################
+
+# GPD Win, Classic and XBox 360 compat modes
+joystick:usb:v11C5p5507*
+joystick:usb:v045Ep028E*
```
Comment (whot) [inline hwdb/70-joystick.hwdb]: vid/pid must be lowercase hex, this won't match otherwise

Comment (hadess) [reply hwdb/70-joystick.hwdb]: Dang it. To be fair, I copied the modalias directly from the udevadm output. I certainly expected the matching not to care :(

Comment (whot) [reply hwdb/70-joystick.hwdb]: it's a weird situation, the modalias is uppercase hex but anything using attr{id/vendor} needs lowercase. We should probably add bits to the parse_hwdp check to warn about that, but...

**Comment by hadess** [comment]
I've made the changes requested. Was there anything other than testing left?

File: hwdb/70-joystick.hwdb [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+# The lookup keys are composed in:
+#   70-joystick.rules
+#
+# Note: The format of the "joystick:" prefix match key is a
+# contract between the rules file and the hardware data, it might
+# change in later revisions to support more or better matches, it
+# is not necessarily expected to be a stable ABI.
+#
+# Match string format:
+# joystick:<subsystem>:v<vid>p<pid>:name:<name>:
```
Comment (whot) [inline hwdb/70-joystick.hwdb]: this should be bustype, not subsystem, I think?

File: rules/70-joystick.rules [hunk 4]
```diff
@@ -0,0 +1,13 @@
+# do not edit this file, it will be overwritten on update
+
+ACTION=="remove", GOTO="joystick_end"
+ENV{ID_INPUT}=="", GOTO="joystick_end"
+ENV{ID_INPUT_JOYSTICK}=="", GOTO="joystick_end"
```
Comment (whot) [inline rules/70-joystick.rules]: nitpick, if ID_INPUT_JOYSTICK is set but not ID_INPUT then that's a bug elsewhere, you can drop the ID_INPUT check

File: rules/70-joystick.rules [hunk 5]
```diff
@@ -0,0 +1,13 @@
...
+KERNEL!="event*", GOTO="joystick_end"
+
+# joystick:<subsystem>:v<vid>p<pid>:name:<name>:*
```
Comment (whot) [inline rules/70-joystick.rules]: s/subsystem/bustype/

File: hwdb/70-joystick.hwdb [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+# or create a bug report on https://github.com/systemd/systemd/issues and
+# include your new rules, a description of the device, and the output of
+#   udevadm info /dev/input/eventXX.
+#
+# Permitted keys:
+#   Specify if a joystick is a built-in one or external:
+#   ID_INPUT_JOYSTICK_INTEGRATION=internal|external
+#
+# If the property is missing, user-space can assume
+# ID_INPUT_JOYSTICK_INTEGRATION=external
```
Comment (whot) [inline hwdb/70-joystick.hwdb]: nitpick: I'd indent that to be on the same as the property description

**Comment by poettering** [comment]
@hadess have you seen @whot's comments?

**Comment by poettering** [comment]
@hadess i figure you removing the needs rework label means your approval. Will merge hence

**Comment by hadess** [comment]
> @hadess i figure you removing the needs rework label means your approval. Will merge hence

I meant it as "it shouldn't need rework now but please review". Let's hope it doesn't break anything :)

**Comment by poettering** [comment]
oh, dang, i didn't realize this was your PR in the first place

**Pull request merged**